dcicutils/sheet_utils.py

```diff
@@ -128,24 +157,128 @@ def set_path_value(cls, datum: Union[List, Dict], path: ParsedHeader, value: Any
             cls.set_path_value(datum[key], more_path, value)
 
 
-class WorkbookManager:
+class AbstractTableSetManager:
+    """
+    The TableSetManager is the spanning class of anything that wants to be able to load a table set,
+    regardless of what it wants to load it from. To do this, it must support a load method
+    that takes a filename and returns the file content in the form:
+        {
+            "Sheet1": [
+                          {...representation of row1 as some kind of dict...},
+                          {...representation of row2 as some kind of dict...}
+                      ],
+            "Sheet2": [...],
+            ...,
+        }
+    Note that at this level of abstraction, we take no position on what form of representation is used
+    for the rows, as long as it is JSON data of some kind. It might be
+         {"col1": "val1", "col2", "val2", ...}
+    or it might be something more structured like
+         {"something": "val1", {"something_else": ["val2"]}}
+    Additionally, the values stored might be altered as well. In particular, the most likely alteration
+    is to turn "123" to 123 or "" to None, though the specifics of whether and how such transformations
+    happen is not constrained by this class.
+    """
+
+    def __init__(self, **kwargs):
+        if kwargs:
+            raise ValueError(f"Got unexpected keywords: {kwargs}")
 
     @classmethod
-    def load_workbook(cls, filename: str):
-        wb = cls(filename)
-        return wb.load_content()
+    def load(cls, filename: str) -> Dict[str, List[AnyJsonData]]:
+        """
+        Reads a filename and returns a dictionary that maps sheet names to rows of dictionary data.
+        For more information, see documentation of AbstractTableSetManager.
+        """
+        raise NotImplementedError(f".load(...) is not implemented for {cls.__name__}.")
```

Comment:
Seems like this should be an [abstract base class](https://docs.python.org/3/library/abc.html)?

Reply:
I don't think I was quite clean enough. I know of such classes from Lisp, but hadn't been tracking Python's adoption of them. As I read the doc, having a method that does something isn't helpful because it's not in the MRO. Here I'm relying on `__init__(self, **kwargs)` for some error handling that must be in the MRO.

I also gotta say that although I do find it useful to name these things in the way that says yeah, it's abstract, I have not found metaclasses to be as useful as I expected them to be. Almost every time I've found a use for metaclasses, there's a just as easy way to make it with regular classes. (Actually, in Python I often use decorators.) Part of the issue is that the uses invariably thwart static analysis, so I end up finding static analysis more important than the other stuff. Maybe such tools would understand this class as a special case because it might have widespread use, but in general metaclasses just make it hard to understand. We have code in `dcicutils.misc_utills` for example that imlements `classproperty` but PyCharm fusses at me because it doesn't understand why I'm using `cls` instead of `self` as an argument.

So I'm gonna pass on that at this time. I'll add a note in the sources for future consideration. If you'd like to convince me otherwise, let's chat interactively. I'm not fixed on this. It's just my, uh, base posture.